beaker_kernel/lib/subkernels/base.py

```diff
@@ -12,9 +26,110 @@ class BaseSubkernel(abc.ABC):
+                    headers={"Authorization": f"token {server_token}"},
+                )
+                if res.status_code == 204:
+                    self.jupyter_id = None
+            except requests.exceptions.HTTPError as err:
+                print(err)
+
+
+class BaseCheckpointableSubkernel(BaseSubkernel):
+    SERIALIZATION_EXTENSION: str = "storage"
+
+    def __init__(self, jupyter_id: str, subkernel_configuration: dict, context):
+        super().__init__(jupyter_id, subkernel_configuration, context)
+        self.checkpoints = None
+        self.storage_prefix = None
+
+    @property
+    def checkpoints_enabled(self):
+        return self.checkpoints is not None
+
+    def start_checkpointing(self):
+        self.checkpoints : list[Checkpoint] = []
+        self.storage_prefix = mkdtemp()
+        makedirs(self.storage_prefix, exist_ok=True)
+    
+    def store_serialization(self, filename: str) -> str:
+        with open(filename, "rb") as file:
+            chunksize = 4 * 1024 * 1024
+            hash = hashlib.new("sha256")
+            while chunk := file.read(chunksize):
+                hash.update(chunk)
+            identifier = hash.hexdigest()
+            new_filename = f"{self.storage_prefix}/{identifier}.{self.SERIALIZATION_EXTENSION}"
+
+        shutil.move(filename, new_filename)
+        return new_filename
+
+    @abc.abstractmethod
+    async def generate_checkpoint_from_state(self) -> Checkpoint:
+        ...
+
+    @abc.abstractmethod
+    async def load_checkpoint(self, checkpoint: Checkpoint):
+        ...
+
+    @action()
+    async def add_checkpoint(self, message):
+        if not self.checkpoints_enabled:
+            self.start_checkpointing()
+        fetched_checkpoint = await self.generate_checkpoint_from_state()
+        checkpoint = {
+            varname: self.store_serialization(filename) for
+            varname, filename in fetched_checkpoint.items()
+        }
+        self.checkpoints.append(checkpoint)
+        return len(self.checkpoints)
+    add_checkpoint._default_payload = "{}"
+   
+    @action()
+    async def rollback(self, message):
+        if not self.checkpoints_enabled:
+            self.start_checkpointing()
+        checkpoint_index = message.content.get("checkpoint_index", None)
+        if checkpoint_index is None:
+            raise RuntimeError("No checkpoint index provided")
+        if checkpoint_index >= len(self.checkpoints):
+            raise IndexError(f"Checkpoint at index {checkpoint_index} does not exist")
+        checkpoint = self.checkpoints[checkpoint_index]
+        await self.load_checkpoint(checkpoint)
+        self.checkpoints = self.checkpoints[:checkpoint_index + 1]
+    
+    rollback._default_payload = "{\n\t\"checkpoint_index\": 0\n}"
+
+    def cleanup(self):
+        if self.checkpoints_enabled:
+            shutil.rmtree(self.storage_prefix, ignore_errors=True)
+            self.checkpoints = []
+        super().cleanup()
```

Comment:
Let's do the super before the cleanup for the checkpoints. We want to ensure that super cleanup will always get called, and as early as possible.